Remote sampling support - AWS X-Ray

bhautikpip opened the pull request:
This PR introduces remote sampling support using AWS X-Ray. This implementation uses 2 X-Ray APIs as mentioned below.

1. getSamplingRules (retrieve customer set rules on AWS X-Ray console)
2. getSamplingTargets (retrieve targets assigned to each client for each rule but API request should also send a snapshot on how many requests are sample, borrowed and matched using each rule)

Reference docs for some of the data structures used in implementation
https://docs.aws.amazon.com/xray/latest/devguide/xray-console-sampling.html#xray-console-config
https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingRules.html
https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingTargets.html

bhautikpip commented:
@anuraaga @MrAlias @jmacd @Aneurysm9 Can you review this PR? I have closed the older one and opened a fresh PR with complete remote sampling support and tried to address most of the concerns/comments we have discussed. Feel free to take a look again. Thanks!

anuraaga commented on samplers/aws/xray/internal/util/rand.go:
```diff
@@ -0,0 +1,92 @@
+var _ rand.Source = (*lockedSource)(nil)
+var _ rand.Source64 = (*lockedSource64)(nil)
+
+type lockedSource struct {
```
Since the rand is only used in a ticker, which is a single thread, don't think you need a locked source

hanyuancheung commented on samplers/aws/xray/fallback_sampler.go:
```diff
@@ -0,0 +1,69 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
```
```suggestion
package xray // import "go.opentelemetry.io/contrib/samplers/aws/xray"
```

hanyuancheung commented on samplers/aws/xray/internal/client.go:
```diff
@@ -0,0 +1,191 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
```
```suggestion
package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
```

hanyuancheung commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+	return minPoll
+}
+
+// generateClientId generates random client ID
```
```suggestion
// generateClientID generates random client ID
```

hanyuancheung commented on samplers/aws/xray/internal/match.go:
```diff
@@ -0,0 +1,71 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
```
```suggestion
package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
```

hanyuancheung commented on samplers/aws/xray/internal/match_test.go:
```diff
@@ -0,0 +1,114 @@
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
```
Invalid blank line could be removed

hanyuancheung commented on samplers/aws/xray/internal/reservoir.go:
```diff
@@ -0,0 +1,79 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
```
```suggestion
package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
```

hanyuancheung commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,205 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
```
```suggestion
package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
```

hanyuancheung commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,205 @@
+	}
+}
+
+// Sample uses sampling targets of a given rule to decide which sampling should be done and returns a SamplingResult.
```
```suggestion
// Sample uses sampling targets of a given rule to decide 
// which sampling should be done and returns a SamplingResult.
```

hanyuancheung commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,205 @@
+	return sd
+}
+
+// appliesTo performs a matching against rule properties to see if a given rule does match with any of the rule set on AWS X-Ray console
```
```suggestion
// appliesTo performs a matching against rule properties to see 
// if a given rule does match with any of the rule set on AWS X-Ray console.
```

hanyuancheung commented on samplers/aws/xray/internal/util/clock.go:
```diff
@@ -0,0 +1,52 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
```
```suggestion
package util // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
```

hanyuancheung commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,175 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
```
```suggestion
package xray // import "go.opentelemetry.io/contrib/samplers/aws/xray"
```

hanyuancheung commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,175 @@
+	"time"
+
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
+
+	"github.com/go-logr/logr"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
```
```suggestion
	"go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"

	"github.com/go-logr/logr"
```

hanyuancheung commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	// Use fallback sampler if manifest is expired or sampling rules does not match against manifest
```
```suggestion
	// Use fallback sampler if manifest is expired or sampling rules 
	// does not match against manifest
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
```
```suggestion
package xray // import "go.opentelemetry.io/contrib/samplers/aws/xray"
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+	defaultPollingInterval = 300
+)
+
+// SamplerOption is a function that sets config on the sampler
```
Wrong annotation. could you please fix it?

hanyuancheung commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+	logger                       logr.Logger
+}
+
+// sets custom proxy endpoint
```
```suggestion
// WithEndpoint sets custom proxy endpoint
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+// sets polling interval for sampling rules
```
```suggestion
// WithSamplingRulesPollingInterval sets polling interval for sampling rules
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+// sets custom logging for remote sampling implementation
```
```suggestion
// WithLogger sets custom logging for remote sampling implementation
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_test.go:
```diff
@@ -0,0 +1,56 @@
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// assert that when manifest is not expired sampling happens with 1 req/sec.
```
```suggestion
// TestShouldSample assert that when manifest is not expired sampling happens with 1 req/sec.
```

hanyuancheung commented on samplers/aws/xray/remote_sampler_test.go:
```diff
@@ -0,0 +1,56 @@
+	assert.Equal(t, sd.Decision, sdktrace.RecordAndSample)
+}
+
+// assert remote sampling description.
```
```suggestion
// TestRemoteSamplerDescription assert remote sampling description.
```

hanyuancheung commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+	return m.refreshedAt < m.Clock.Now().Unix()-manifestTTL
+}
+
+// MatchAgainstManifestRules returns a Rule and boolean flag set as true if rule has been match against span attributes, otherwise nil and false
```
```suggestion
// MatchAgainstManifestRules returns a Rule and boolean flag set as true 
// if rule has been match against span attributes, otherwise nil and false.
```

hanyuancheung reviewed (changes requested):
PTAL at the suggestion.

bhautikpip commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+
+// MatchAgainstManifestRules returns a Rule and boolean flag set as true if rule has been match against span attributes, otherwise nil and false
+func (m *Manifest) MatchAgainstManifestRules(parameters sdktrace.SamplingParameters, serviceName string, cloudPlatform string) (*Rule, bool, error) {
+	m.mu.RLock()
```
I think we need read lock here reason being when we do this type of assignment `rules := m.Rules` even then the underlying copy would still remain the same meaning `&m.Rules[0]` and `&rules[0]` would still point to the same address. So I think adding read lock here for the life cycle of a function makes the code more safer because manifest might get updated when runtime is executing this logic.

anuraaga replied:
In that case something changed in the refactoring since we meant for the entire manifest to be read once for sampling, and be immutable to not need a lock for the entire sampling decision. IIRC we had gotten to that in the previous iteration - my reading of the code was that in this iteration, `m.Rules` is the equivalent of the previous `s.manifest` which we were treating as immutable

bhautikpip replied:
Yeah, you're right `m.Rules` is immutable but we still need read locks here because we are iterating over rules to see if there is a match and if there is a match then we sample. Meanwhile other thread can update the manifest rules/targets right? So, in order to synchronize that we need read locks here. Also, we can't do `rules := m.Rules` which basically means we are making a shallow copy so we are iterating over the same memory addresses. 

To explain more:

Main thread ---> reads manifest (for sampling and makes a sampling decision) (read op)
     |
     |
     - - - - - - -> updates manifest in one refresh (write op)
     (second thread)

anuraaga replied:
`m.Rules is immutable` and `update the manifest rules` seem to be a contradiction. If the rules are immutable, they can't be updated by another thread, the reason we need a lock to read `m.Rules` is memory visibility, but after taking a reference, if the object is immutable than there would be no further need for locking

bhautikpip replied:
yep we should be okay to follow this suggestion

bhautikpip commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+
+	// deep copy centralized manifest object to temporary manifest to avoid thread safety issue
+	m.mu.RLock()
+	err = copier.CopyWithOption(&manifest, m, copier.Option{IgnoreEmpty: false, DeepCopy: true})
```
Yep had a chat with @MrAlias regarding the same. He suggested that we might need to use `reflect` package to build out something like this. I found this open source project which was addressing our need so used it instead writing our own logic. I could not find a good way to do this other than using `reflect` package in go. Just curious, why should we avoid using `reflect` package here? I would appreciate any other better suggestions :)

anuraaga replied:
I think hand writing the copying would be fairly straightforward without needing reflect or reflection. I'd generally avoid reflection in production code unless absolutely required to make the code easier to reason about since some times reflection copies can copy in unintended ways causing bugs, but if maintainers are OK with this library than it's fine I guess

bhautikpip replied:
yep will wait on maintainers feedback on using this library.

bhautikpip commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.refreshedAt < m.Clock.Now().Unix()-manifestTTL
```
Just curious, any particular reason on why shouldn't we use Unix timestamp? Basically the [request and response](https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingTargets.html) of `getSamplingTargets` API consist unix timestamp so to avoid having to do back and forth conversion between Unix timestamp and `time.Time` decided to use `Unix` everywhere.

anuraaga replied:
A leap second or similar can pass and cause timestamp differences to not be quite accurate so we should avoid Unix as much as possible. This means that the unix timestamps in the API response get converted to `time` (monotonic time) ASAP and not using unix time anymore, and also `Interval` isn't unix time so wouldn't have that issue either.

anuraaga commented on samplers/aws/xray/internal/reservoir.go:
```diff
@@ -0,0 +1,81 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Considering copying the code in from here to not have sampling decisions aligned on epoch second boundaries like this implementation. The alignment has some potential for thundering herd problems so is somewhat operationally unsafe

https://github.com/jaegertracing/jaeger-client-go/blob/master/utils/rate_limiter.go

anuraaga commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,213 @@
+
+// Rule represents a sampling rule which contains rule properties and reservoir which keeps tracks of sampling statistics of a rule
+type Rule struct {
+	samplingStatistics samplingStatistics
```
Believe this needs to be a pointer

bhautikpip replied:
Nah it does not necessarily have to be a pointer. I have tested the code and works as expected with this non-pointer declaration.

anuraaga replied:
The deep copy will copy the struct so I can't see how the statistics could be updated during refresh if this isn't a pointer. If you think it is, please explain the flow

bhautikpip replied:
yep I changed it to pointer. If we don't change it to pointer then values get updated when we assign deep copy back to centralized object but making this struct pointer makes more sense since values would get updated even when deep copy logic gets executed. So changed it to make this a pointer.

anuraaga commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,213 @@
+// samplingStatisticsDocument. It also resets statistics counters.
+func (r *Rule) snapshot(now time.Time) *samplingStatisticsDocument {
+	name := r.ruleProperties.RuleName
+	requests, sampled, borrowed := r.samplingStatistics.matchedRequests, r.samplingStatistics.sampledRequests, r.samplingStatistics.borrowedRequests
```
I remember commenting previously that we want to use `atomic.swap` for these get/resets

anuraaga commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,219 @@
+func (r *Rule) snapshot(now time.Time) *samplingStatisticsDocument {
+	name := r.ruleProperties.RuleName
+
+	matchedRequests := atomic.LoadInt64(&r.samplingStatistics.matchedRequests)
```
Sorry if my comment was confusing, I mean something like

`matchedRequests := atomic.SwapInt64(&matchedRequests, int64(0))`

anuraaga commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,351 @@
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	now := time.Unix(m.Clock.Now().Unix()-manifestTTL, 0)
```
Don't think you should need `.Unix` here since we are just using relative times

anuraaga commented on samplers/aws/xray/internal/reservoir.go:
```diff
@@ -0,0 +1,116 @@
+
+	currentTime := now
+
+	if currentTime.Unix() == r.borrowTick.Unix() {
```
Borrowing is also just a 1QPS rate limit, let's remove the epoch-based borrowing and use the same logic as the other rate limiting, similar to what the Java implementation does. Perhaps it will mean extracting a `rate_limiter` type. It's not as dangerous as the normal rate limiting but any epoch based logic has some chance of thundering herd

Aneurysm9 commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,351 @@
+}
+
+// minimumPollingInterval finds the minimum interval amongst all the targets
+func (m *Manifest) minimumPollingInterval() (minPoll time.Duration) {
+	minPoll = time.Duration(0)
+	for _, rules := range m.Rules {
+		if minPoll == 0 {
+			minPoll = rules.reservoir.interval
+		} else {
+			if minPoll >= rules.reservoir.interval {
+				minPoll = rules.reservoir.interval
+			}
+		}
+	}
+
+	return minPoll * time.Second
+}
```
```suggestion
func (m *Manifest) minimumPollingInterval() time.Duration {
	minPoll = math.MaxInt
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}

	return minPoll * time.Second
}
```

This could still end up at zero, so not sure having two separate branches does much good.  If you want to return `0` in the case of an empty list then do that as a check up-front:

```go
if len(m.Rules) == 0 {
  return time.Duration(0)
}
```

bhautikpip replied:
Okay I will update the logic like below. if we initialize `minPoll = math.MaxInt` then caller has to check 2 conditions `minPoll != math.MaxInt` and `minPoll != 0` because we don't want to update `m.SamplingTargetsPollingInterval` with zero or `math.MaxInt` in that case we would rather not update. So if we initialize with 0 then we can achieve this by only checking `minPoll != 0` and we can update the `m.SamplingTargetsPollingInterval`

```
	minPoll = time.Duration(0)
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
```

Aneurysm9 replied:
```go
	minPoll = time.Duration(0)
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
```

This might as well be:

```go
return time.Duration(0)
```

If, instead, you use this:

```go
func (m *Manifest) minimumPollingInterval() time.Duration {
	if len(m.Rules) == 0 {
		return time.Duration(0)
	}

	minPoll = math.MaxInt
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
}
```

It will only return `time.Duration(0)` if there are no rules and will never return `math.MaxInt` as there will be at least one rule and that rule's interval will be less than `math.MaxInt` in any sane universe.  If it isn't, then that truly is the minimum interval and so be it.

Aneurysm9 commented on samplers/aws/xray/internal/reservoir.go:
```diff
@@ -0,0 +1,112 @@
+		if elapsedTime.Seconds() > 1.0 {
+			r.quotaBalance += 1.0
+		} else {
+			r.quotaBalance += elapsedTime.Seconds() * 1
```
```suggestion
			r.quotaBalance += elapsedTime.Seconds()
```

Aneurysm9 commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,214 @@
+// attributeMatching performs a match on attributes set by users on AWS X-Ray console
+func (r *Rule) attributeMatching(parameters sdktrace.SamplingParameters) (match bool, err error) {
+	match = false
+	unmatchedCounter := 0
+	if len(r.ruleProperties.Attributes) > 0 {
+		for key, value := range r.ruleProperties.Attributes {
+			for _, attrs := range parameters.Attributes {
```
```suggestion
	if len(r.ruleProperties.Attributes) > 0 {
		for key, value := range r.ruleProperties.Attributes {
			unmatchedCounter := 0
			for _, attrs := range parameters.Attributes {
```

Aneurysm9 commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,214 @@
+			if unmatchedCounter == len(parameters.Attributes) {
+				return false, nil
+			}
+			unmatchedCounter = 0
```
```suggestion
```

Aneurysm9 commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,214 @@
+
+// attributeMatching performs a match on attributes set by users on AWS X-Ray console
+func (r *Rule) attributeMatching(parameters sdktrace.SamplingParameters) (match bool, err error) {
+	match = false
```
```suggestion
```

This is the zero value.  Though I'm not sure a named return value is needed here, so maybe remove that and instead:

```suggestion
	match := false
```

Aneurysm9 commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,189 @@
+// ShouldSample matches span attributes with retrieved sampling rules and perform sampling,
+// if rules does not match or manifest is expired then use fallback sampling.
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	if !rs.manifest.Expired() {
```
```suggestion
	if rs.manifest.Expired() {
		// Use fallback sampler if manifest is expired
		rs.logger.V(5).Info("manifest is expired so using fallback sampling strategy")
		return rs.fallbackSampler.ShouldSample(parameters)
	}
```

Invert this condition and outdent the remainder, it'll be easier to follow.

Aneurysm9 replied:
See the next comment.  The current code is doing double duty and should be split into two cases.  This immediate fallback if the manifest is expired and the later ultimate fallback if there are no rule matches.  They are separate cases and should be handled separately.  Doing so has the added benefit of making the logic flow clearer.

Aneurysm9 commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,189 @@
+		}
+	}
+
+	// Use fallback sampler if manifest is expired
+	// or sampling rules does not match against manifest
+	rs.logger.V(5).Info("span attributes does not match to the sampling rules or manifest is expired so using fallback sampling strategy")
+	return rs.fallbackSampler.ShouldSample(parameters)
```
```suggestion
	// Use fallback sampler if
	// sampling rules does not match against manifest
	rs.logger.V(5).Info("span attributes does not match to the sampling rules so using fallback sampling strategy")
	return rs.fallbackSampler.ShouldSample(parameters)
```

bhautikpip replied:
We should keep this right? As I have mentioned above fallback sampling can be called in either of the 2 cases.

Aneurysm9 commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+func (rs *remoteSampler) getDescription() string {
+	return "remote sampling with AWS X-Ray"
```
Any reason not to just inline this in `Description()`?

bhautikpip replied:
Initially it was inlined only but @anuraaga suggested to use `getDescription()` method probably to have implementation parity with Java implementation. I don't have any preference on this. I'm fine with or without the method.

Aneurysm9 replied:
Unless there is logic to be performed here, just inline it.  It can always be changed later if we decide there is something interesting to do to build the description.

anuraaga replied:
Ah the suggestion was to make the description itself follow the pattern of the java implementation, not the method name

bhautikpip replied:
oh okay. I might have misread the comment then. Yeah we should be okay to inline this then.

Aneurysm9 commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,189 @@
+// startPoller starts the rule and target poller in a single go routine which runs periodically
+// to refresh manifest and targets.
+func (rs *remoteSampler) startPoller(ctx context.Context) {
+	go func() {
```
```suggestion
```

Consider removing this and invoking as `go rs.startPoller(ctx)` instead.

Aneurysm9 commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,109 @@
+	defaultPollingInterval = 300
+)
+
+// Option is a function that sets config on the sampler.
+type Option func(options *config)
```
```suggestion
// Option sets configuration on the sampler.
type Option interface {
	apply(*config)
}
```

See the [style guide for options](https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#option).

You can make an `optionFunc` implementation to deal with simple options:

```go
type optionFunc func(*config)

func (f optionFunc) apply(cfg *config) {
	f(cfg)
}
```

bhautikpip commented on samplers/aws/xray/internal/client.go:
```diff
@@ -0,0 +1,200 @@
+
+// getSamplingRules calls the collector(aws proxy enabled) for sampling rules.
+func (c *xrayClient) getSamplingRules(ctx context.Context) (*getSamplingRulesOutput, error) {
+	samplingRulesInput, err := json.Marshal(getSamplingRulesInput{})
```
Yeah `nextToken` field is never used and the API expects nil value of that but I have created this struct to use more like a place holder struct for this. Maybe in future if API changes and if we want to leverage this value inside this implementation then we can without introducing anything at that time. I don't see any overhead on keeping it in a struct like this. I think the API without input params looks more confusing so maybe we can just keep it like this but happy to hear your thoughts if you think otherwise.

bhautikpip commented on samplers/aws/xray/internal/client.go:
```diff
@@ -0,0 +1,200 @@
+
+// newClient returns an HTTP client with proxy endpoint.
+func newClient(addr string) (client *xrayClient, err error) {
+	endpoint := "http://" + addr
```
I think we might need `http` (local testing) and `https` both so I was just thinking rather than taking endpoint from user we can just directly take host like `http://localhost:2000` or `https://localhost:2000`. Also, I'm not super sure that `awsxrayproxy` supports HTTPS, @anuraaga any comments on that? We can validate the host in `validateConfig` method. @Aneurysm9 does this sound good to you?

bhautikpip commented:
@hanyuancheung @MrAlias @jmacd Can you also take a look? planning to merge this ASAP to include this in next release if possible.

bhautikpip commented:
@Aneurysm9 addressed your comments. Feel free to take a look at it again and suggest if I missed anything :)

Aneurysm9 commented:
@hanyuancheung if all of your comments have been addressed can you please approve and release the hold on this PR?

MadVikingGod commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,186 @@
+// to refresh manifest and targets.
+func (rs *remoteSampler) startPoller(ctx context.Context) {
+	// jitter = 5s, default 300 seconds
+	rulesTicker := util.NewTicker(rs.samplingRulesPollingInterval, 5*time.Second)
+
+	// jitter = 100ms, default 10 seconds
+	targetTicker := util.NewTicker(rs.manifest.SamplingTargetsPollingInterval, 100*time.Millisecond)
```
Nit:  These tickers are never closed, so they will "leak" just like a [`time.Tick()`](https://pkg.go.dev/time#Tick)

MadVikingGod commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	for index := range m.Rules {
+		if m.Rules[index].ruleProperties.RuleName == *t.RuleName {
+			m.Rules[index].reservoir.refreshedAt = m.clock.Now()
+
+			// Update non-optional attributes from response
+			m.Rules[index].ruleProperties.FixedRate = *t.FixedRate
+
+			// Update optional attributes from response
+			if t.ReservoirQuota != nil {
+				m.Rules[index].reservoir.quota = *t.ReservoirQuota
+			}
+			if t.ReservoirQuotaTTL != nil {
+				m.Rules[index].reservoir.expiresAt = time.Unix(int64(*t.ReservoirQuotaTTL), 0)
+			}
+			if t.Interval != nil {
+				m.Rules[index].reservoir.interval = time.Duration(*t.Interval)
+			}
+		}
```
You need to take the Rule's lock to update these fields, or Rule needs an update method that captures that lock.

bhautikpip replied:
Since we're doing deep copy [here ](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1859/files/0cf3224c740202e79bd33649f307941f9954ee55#diff-d909e7bd591a0ccb2cedd3b1602921ca5ca1612d282189172fb3d666fffdcfd0R127) we don't need lock here while updating reservoir values. While [updating the ruleProperties](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1859/files/0cf3224c740202e79bd33649f307941f9954ee55#diff-d909e7bd591a0ccb2cedd3b1602921ca5ca1612d282189172fb3d666fffdcfd0R174) we are creating manifest every time so don't think we need lock there either.

MadVikingGod commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,214 @@
+	atomic.AddInt64(&r.samplingStatistics.matchedRequests, int64(1))
+
+	// fallback sampling logic if quota for a given rule is expired
+	if r.reservoir.expired(now) {
```
How does one get a Manifest that isn't expired that produces a rule with an expired reservoir?  This seems to split the fallback logic into multiple places.

Could we rid ourselves of this fallback code if `Manifest.MatchAgainstManifestRules` also checked if the reservoir (or the Rule) is expired and errored there?

bhautikpip replied:
manifest update time simply refers to the time when we refreshed all the rules. `reservoir. expired(now)` checks the TTL for a particular rule returned by this [API response](https://docs.aws.amazon.com/xray/latest/api/API_SamplingTargetDocument.html#xray-Type-SamplingTargetDocument-ReservoirQuotaTTL). So non-expired manifest can contain expired rules based on the `reservoirQuotaTTL` value. Yeah I agree we are splitting the fallback logic but we actually need to cover all these cases. Nah I think if rule is expired then we need to sample 1 req/sec and 5% additional request during that second so even if we migrate this to the method you suggested we need this logic there and don't see any benefit of splitting this code since I feel this looks more readable.

MadVikingGod commented on samplers/aws/xray/internal/util/timer.go:
```diff
@@ -0,0 +1,45 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
```
This kind of package name is an anti-pattern.  As much as possible these should be either in the internal package or something like `ticker`.

bhautikpip replied:
That's a good call out. I can move these files to internal package.

MadVikingGod reviewed (commented):
This isn't a full review, but there is some actionable feedback.

bhautikpip commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,186 @@
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, serviceName string, cloudPlatform string, opts ...Option) (sdktrace.Sampler, error) {
```
Yeah  think this is the expected behavior. That's a good point. I will document this behavior.

MadVikingGod replied:
Was this documented anywhere?

bhautikpip replied:
documented it here: https://github.com/aws-otel/aws-otel.github.io/pull/269/files#diff-ad596b6a242f203456c990cac4fb6d6e94b621bae7abbe15d6f5ca79830faba7R194

bhautikpip commented:
@MadVikingGod Can you take another look if your comments have been successfully addressed or any more suggestions on this PR? I'm planning to include this change in the next release so would really appreciate a quick review :) Thanks :)

MadVikingGod commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,352 @@
+	Rules                          []Rule
+	SamplingTargetsPollingInterval time.Duration
+	refreshedAt                    time.Time
+	xrayClient                     *xrayClient
```
Nit:  This should probably be an interface, so you don't have to create an http server etc to test your manifest code.  It could be an internal interface that the xray client Already meets.

bhautikpip replied:
not sure if I follow you on this. I think we need http server to mimic `getSamplingRules` and `getSamplingTargets` API response.

MadVikingGod commented:
I had trouble decerning if the manifest type you have is thread-safe.  Two things you could do is:
1. Put all of your lock logic at the same level, all in the Exported functions, or use your function names to differentiate what must be used within a lock.
2. Build some race tests around your Exported functions.

bhautikpip commented on samplers/aws/xray/internal/reservoir.go:
```diff
@@ -0,0 +1,112 @@
+	// stores reservoir ticks
+	lastTick time.Time
+
+	mu *sync.RWMutex
```
I think to avoid having this warning - https://stackoverflow.com/questions/37052757/how-to-get-around-assignment-copies-lock-value-to-tr-net-http-transport-contain

MrAlias replied:
If you work with a `*reservoir` instead of a `reservoir` (similar to [the linked answer](https://stackoverflow.com/a/37052960/3614233)), you should not have an issue with copying the lock. All of your method receivers are pointers so it seems like that might be the better approach, no?

bhautikpip replied:
I think now that I remember the main motivation of making `reservoir` struct non pointer was to get a new address space when we deep copy the manifest object so don't think making `reservoir` pointer would help in this case. Also, since  we are doing deep copy I'm not sure if it's a good idea to keep this non pointer since if reservoir's `mu` gets updated we would be overriding stale lock values.

bhautikpip replied:
I'm not sure this is true. we're applying read lock and that is also for a very brief period of time. This is how the code flow will look like. In below flow we're assuming mutex is a pointer.

Code applies read lock  -> Deep copy is performed (new copy has a read locked mutex) -> code release the lock (both copies are available for updates)

While we're doing a deep copy there won't be any updates anyways. so no need to worry about that even a very brief period of time. 

Yeah, I still don't think custom copy is needed here since there are already open sourced libraries available. I don't think deep copy is issue here anyways. Also, do you see any potential issues with mutex being a pointer? Is it not considered a good practice in golang?

MrAlias replied:
There is a reason govet warns about copying locks, and it is for this very reason. You make the copy and the lock now applies to 2 (or more) values.

What are you doing if you lock the original lock? You have locked both the original and the copy (or copies). Should they still be updated? According to their lock status they should not.

Should the original be locked if the copy is locked? As it is currently designed it will be.

When you make a copy of the original it needs to produce a value that has its own lock. If you can find a way to achieve this without a custom copy method that sounds good, but otherwise you need to make a custom copy method to ensure this.

bhautikpip commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,352 @@
+
+	// consume unprocessed statistics messages
+	for _, s := range targets.UnprocessedStatistics {
+		m.logger.V(5).Info(
```
Technically unprocessed statistics is something that is returned by X-Ray service as part of the API response  and not anything to have to do with client side. I think this is just informational message so don't think we should flagged it as error log.

bhautikpip commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,352 @@
+// snapshots takes a snapshot of sampling statistics from all rules, resetting
+// statistics counters in the process.
+func (m *Manifest) snapshots() ([]*samplingStatisticsDocument, error) {
+	statistics := make([]*samplingStatisticsDocument, 0, len(m.Rules)+1)
```
removed it.

bhautikpip commented on samplers/aws/xray/internal/manifest_test.go:
```diff
@@ -0,0 +1,1592 @@
+
+// assert that refresh manifest targets successfully updates samplingTargetsPollingInterval.
+func TestRefreshManifestTargets_PollIntervalUpdateTest(t *testing.T) {
+	// RuleName is missing from r2
```
removed.

bhautikpip commented on samplers/aws/xray/internal/reservoir_test.go:
```diff
@@ -0,0 +1,259 @@
+	r.refreshQuotaBalance(clock.now(), false)
+
+	// assert that if quotaBalance exceeds capacity then total capacity would be new quotaBalance
+	assert.Equal(t, r.quotaBalance, 1*r.capacity)
```
removed.

bhautikpip commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,186 @@
+
+// Description returns description of the sampler being used.
+func (rs *remoteSampler) Description() string {
+	return "AwsXrayRemoteSampler{remote sampling with AWS X-Ray}"
```
I think keeping the description is meant to have parity between various language implementation.

bhautikpip commented on samplers/aws/xray/rand.go:
```diff
@@ -0,0 +1,39 @@
+}
+
+func newGlobalRand() *rand.Rand {
+	src := rand.NewSource(newSeed())
```
okay made it a package level var `var seed = newSeed()`

bhautikpip commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,186 @@
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, serviceName string, cloudPlatform string, opts ...Option) (sdktrace.Sampler, error) {
```
equivalent to AWS X-Ray `serviceType` from AWS X-Ray [documentation](https://docs.aws.amazon.com/xray/latest/devguide/xray-console-sampling.html)

MrAlias replied:
> equivalent to AWS X-Ray `serviceType` from AWS X-Ray [documentation](https://docs.aws.amazon.com/xray/latest/devguide/xray-console-sampling.html)

Please include that in the function documentation.

bhautikpip replied:
will add this documentation: https://aws-otel.github.io/docs/getting-started/remote-sampling#otel-remote-sampling-implementation-caveats

MrAlias replied:
Please add the documentation to this code base.

bhautikpip replied:
Yes added the documentation to the `NewRemoteSampler` API.

MrAlias replied:
I think adding a link to external docs is helpful, but can we update the `NewRemoteSampler` docs to directly state what value should be passed for`cloudPlatform` and maybe even `serviceName`? Something similar to what was done explaining the passed `ctx`. That way a user does not have to go track down information at another location.

bhautikpip replied:
gotcha. will update this!

bhautikpip commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,186 @@
+	r, match, err := rs.manifest.MatchAgainstManifestRules(parameters, rs.serviceName, rs.cloudPlatform)
+	if err != nil {
+		rs.logger.Error(err, "regexp matching error while matching span and resource attributes")
+		return sdktrace.SamplingResult{}
```
I don't think so because fallback sampler should be used in 2 cases.

1. when manifest is expired
2. when there is no rule match

This case indicates that we're not able to do the rule match due to some error so we should return the error rather then continue with sampling decision.

MrAlias replied:
Sure, but you are returning a sampling result. That result will be used as is, the caller has no concept of the error you encountered and is going to drop the span based on the returned value. That does not seem ideal, especially since we have something more intelligent to fallback to.

bhautikpip replied:
Yeah I'm not sure if we should use fallback sampler here? I feel that is kind of misrepresentation of the case when there is an error in matching logic.  `return sdktrace.SamplingResult{}` would return SamplingDecision as `DROP`?

bhautikpip commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,127 @@
+
+func validateConfig(cfg *config) (err error) {
+	// check endpoint follows certain format
+	endpointHostSplit := strings.Split(cfg.endpoint.Host, ":")
+
+	if len(endpointHostSplit) > 2 {
+		return fmt.Errorf("config validation error: expected endpoint host format is hostname:port")
```
I think the reason being `getSamplingRules` and `getSamplingtargets` API call would be done by [`awsproxy`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/awsproxy) so we would need the correct url with port to make that API call. Not sure url without port would work.

MrAlias replied:
Is there no case the user would use their own proxy listening on port 80 or 443?

bhautikpip commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,127 @@
+
+	hostName := endpointHostSplit[0]
+
+	// validate host name
+	r, err := regexp.Compile("[^A-Za-z0-9.]")
+	if err != nil {
+		return err
+	}
```
`url.Parse` would not throw error on URLs like `url.Parse("http://127.0.0.1@:2000")`. This check helps to avoid those cases.

MrAlias replied:
`"http://127.0.0.1@:2000"` is a valid URL. Given connection errors are communicated to the user this type of error will already be visible. Restricting users to use a subset of URLs has the potential to limit usefulness to some users. Assuming the user will ensure the value they are passing is valid according to their needs will not impose such a restriction and seems fair given ultimate communication of errors.

bhautikpip replied:
Fair. Probably will remove most validation logic since looks like with the usage of `url.Parse` we don't need this.

bhautikpip commented:
> I had trouble decerning if the manifest type you have is thread-safe. Two things you could do is:
> 
> 1. Put all of your lock logic at the same level, all in the Exported functions, or use your function names to differentiate what must be used within a lock.
> 2. Build some race tests around your Exported functions.

Sure I think I can modify the function names to have suffix like `Locked` to better understand the code. I can also write some race tests around exported functions.

bhautikpip commented on samplers/aws/xray/internal/manifest_test.go:
```diff
@@ -0,0 +1,1592 @@
+		nowTime: 1500000000,
+	}
+
+	// sampling target received from centralized sampling backend
+	rate := 0.05
+	quota := float64(10)
+	ttl := float64(1500000060)
+	name := "r1"
+
+	st := samplingTargetDocument{
+		FixedRate:         &rate,
+		ReservoirQuota:    &quota,
+		ReservoirQuotaTTL: &ttl,
+		RuleName:          &name,
+	}
```
I think in some tests we won't need all the values. For example `TestUpdateReservoirMissingFixedRate` we don't need `FixedRate` so containing inside a function might not help. Also, we can populate it like a normal data structure and we're functionally not doing anything so I think it is okay to keep it here.

MrAlias replied:
There is declaration of field values, indirection of those values, and assignment of the pointers to the actual field being done over 10 lines. This pattern is repeated in many places.

Abstracting this to a function call, one where you can pass zero values for  unused fields, would amortize the code space and centralize any future change needed.

bhautikpip replied:
okay will update this.

bhautikpip replied:
I think I have updated this. Please mark this as resolved if there are no further action.

bhautikpip commented:
@MrAlias @MadVikingGod I'm planning to add race test but other than this I have addressed most your comments. Let me know if you have any suggestions/examples for race tests.

bhautikpip commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,214 @@
+		}
+	}
+
+	return attributeMatcher &&
+		serviceNameMatcher &&
+		serviceTypeMatcher &&
+		HTTPMethodMatcher &&
+		HTTPHostMatcher &&
+		HTTPURLPathMatcher, nil
```
good call. we can exit early here.

MrAlias commented on samplers/aws/xray/go.mod:
```diff
@@ -0,0 +1,14 @@
+	github.com/go-logr/stdr v1.2.2
+	github.com/jinzhu/copier v0.3.5
+	github.com/stretchr/testify v1.7.0
+	go.opentelemetry.io/otel v1.4.0
+	go.opentelemetry.io/otel/sdk v1.4.0
+	go.opentelemetry.io/otel/trace v1.4.0
```
```suggestion
	go.opentelemetry.io/otel v1.6.0
	go.opentelemetry.io/otel/sdk v1.6.0
	go.opentelemetry.io/otel/trace v1.6.0
```

MrAlias commented on samplers/aws/xray/internal/client_test.go:
```diff
@@ -0,0 +1,293 @@
+	"github.com/stretchr/testify/require"
+)
+
+func createTestClient(testServerURL string) (*xrayClient, error) {
+	u, err := url.Parse(testServerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClient(*u)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
```
```suggestion
func createTestClient(t *testing.T, body []byte) *xrayClient {
	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, _ *http.Request) {
		_, err := res.Write(body)
		require.NoError(t, err)
	}))
	t.Cleanup(testServer.Close)
	
	u, err := url.Parse(testServer.URL)
	require.NoError(t, err)

	client, err := newClient(*u)
	require.NoError(t, err)

	return client
}
```

MrAlias commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,359 @@
+	return statistics, nil
+}
+
+// sort sorts the Rules of m array first by priority and then by name.
```
```suggestion
// sort sorts the Rules of m first by priority and then by name.
```

MrAlias commented on samplers/aws/xray/internal/manifest_test.go:
```diff
@@ -0,0 +1,1574 @@
+	err = m.RefreshManifestRules(ctx)
+	require.NoError(t, err)
+
+	assert.Len(t, m.Rules, 1)
```
```suggestion
	require.Len(t, m.Rules, 1)
```

MrAlias commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+// ShouldSample matches span attributes with retrieved sampling rules and returns a sampling result.
+// If the sampling parameter to not match or the manifest is expired then the fallback sampler is used.
```
My bad.

```suggestion
// If the sampling parameters do not match or the manifest is expired then the fallback sampler is used.
```

MrAlias commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+// startPoller starts the rule and target poller in a single go routine which runs periodically
+// to the refresh manifest and targets.
```
```suggestion
// to refresh the manifest and targets.
```

MrAlias commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// WithEndpoint sets custom proxy endpoint.
+// default endpoint being used is http://127.0.0.1:2000.
```
```suggestion
// If this option is not provided the default endpoint used will be http://127.0.0.1:2000.
```

MrAlias commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// WithSamplingRulesPollingInterval sets polling interval for sampling rules.
+// default samplingRulesPollingInterval being used is 300 seconds.
```
```suggestion
// If this option is not provided the default samplingRulesPollingInterval used will be 300 seconds.
```

MrAlias commented on samplers/aws/xray/remote_sampler_config.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// WithLogger sets custom logging for remote sampling implementation.
+// default logger being used is go-logr/stdr (https://github.com/go-logr/stdr).
```
```suggestion
// If this option is not provided the default logger used will be go-logr/stdr (https://github.com/go-logr/stdr).
```

bhautikpip commented:
While running the `make precommit` I have started receiving this error. There is some checksum mis match. Are we planning to fix this issue as part of the other PR ? 

```
bhautip@38f9d362ae67 opentelemetry-go-contrib % make precommit     
cd ./tools && \
        go build -o /Users/bhautip/Documents/lambda/opentelemetry-go-contrib/.tools/dbotconf go.opentelemetry.io/build-tools/dbotconf
verifying github.com/blizzy78/varnamelen@v0.6.1/go.mod: checksum mismatch
        downloaded: h1:mGBHm+Uo4e8JnZEKHRoZgVEOQdSBdQfY/x+k4NAXBWA=
        go.sum:     h1:zy2Eic4qWqjrxa60jG34cfL0VXcSwzUrIx68eJPb4Q8=

```

Aneurysm9 commented:
That version [should be retracted](https://github.com/blizzy78/varnamelen/blob/master/go.mod#L12).  Not sure why it is being used.  You may be able to work around it temporarily by putting an `exclude` directive for it in `tools/go.mod`.

MrAlias commented on samplers/aws/xray/internal/manifest.go:
```diff
@@ -0,0 +1,387 @@
+	// Deep copy manifest object.
+	manifest := m.deepCopy()
+
+	// Generate sampling statistics based on the data in temporary manifest
```
```suggestion
	// Generate sampling statistics based on the data in temporary manifest.
```

MrAlias commented:
One final unresolved issue as far as I see: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1859#discussion_r836842309

Otherwise this is looking good :+1:

MrAlias commented on samplers/aws/xray/remote_sampler.go:
```diff
@@ -53,6 +53,8 @@ var _ sdktrace.Sampler = (*remoteSampler)(nil)
 // based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
 // sampling rules and sampling targets.
 // NOTE: ctx passed in NewRemoteSampler API is being used in background go routine. Cancellation to this context can kill the background go routine.
+// serviceName refers to the name of the service equivalent to set by customers on AWS X-Ray console when adding sampling rules and
+// cloudPlatform refers to the environment service is running on (EC2, EKS etc).
```
```suggestion
// serviceName refers to the name of the service equivalent to the one set in the AWS X-Ray console when adding sampling rules and
// cloudPlatform refers to the cloud platform the service is running on ("ec2", "ecs", "eks", "lambda", etc).
```

MrAlias commented on samplers/aws/xray/internal/rule.go:
```diff
@@ -0,0 +1,249 @@
+		}
+	}
+
+	return attributeMatcher && serviceNameMatcher && serviceTypeMatcher && HTTPMethodMatcher && HTTPHostMatcher && HTTPURLPathMatcher, nil
```
This appears to have been reverted to the original form. Why not keep it returning `true` directly?

bhautikpip replied:
yeah I think this would be better for readability and aligns more with what we want to do. We want to return true if all of the matchers are true.

MrAlias replied:
> yeah I think this would be better for readability and aligns more with what we want to do. We want to return true if all of the matchers are true.

Which we do based on the all the code above? Why check those values twice?

I'm not sure it is more reabable if I'm confused. I expect others to read the redundant code and question if there is something more they are missing.

bhautikpip replied:
okay fair. I can change it back to `true`

bhautikpip commented:
Also, not sure why linter check is failing. I'm able to run it on my local. Maybe re-running it would be helpful since it's some timeout issue.

MrAlias commented:
@MadVikingGod PTAL

bhautikpip commented:
@MadVikingGod Can you take a look again to see if you have further comments on this or this is good to merge?

Aneurysm9 commented:
I spoke with @MadVikingGod and he's good with this proceeding.  He will ask for follow-up PRs if he has any concerns that need to be addressed.
